ugs-core/src/com/willwinder/universalgcodesender/MacroHelper.java

```diff
@@ -79,13 +82,14 @@ public class MacroHelper {
      * @param str
      * @param backend
      */
-    public static void executeCustomGcode(final String str, BackendAPI backend) throws Exception {
+    public static void executeCustomGcode(final String[] str, BackendAPI backend) throws Exception {
         if (str == null) {
             return;
         }
-        String command = MacroHelper.substituteValues(str, backend);
-        command = command.replaceAll("(\\r\\n|\\n\\r|\\r|\\n)", "");
-        final String[] parts = command.split(";");
```

Comment:
Macros does support multiline today with a `;`-separator. It looks like you've removed that behaviour which will impact existing macros.

I think it is better to use `\n` instead of `;` like you have suggested. But instead of just dropping the old behaviour you must do some sort of conversion. As `;` is commonly used as a comment prefix, this new conversion should probably only be done once.

May I suggest adding a version to the `Macro`-class. If it doesn't exiss you should treat `;` as newlines and convert. If it has a version tag it should expect `;` as gcode comments and `\n` as newlines.

And by doing this, you don't have to change the type from `String gcode` to `String[] gcode`. And you probably won't need the MacroSerializer and MacroDeserializer.